File: Comparison/go/CRUD/crud.go
package main

import (
	"database/sql"
	"log"
	"net/http"

	"github.com/gin-gonic/gin"
	_ "github.com/go-sql-driver/mysql"
)

// Menu represents a model for menus
type Menu struct {
	ID          int    `json:"id"`
	Name        string `json:"name"`
	URL         string `json:"url"`
	Description string `json:"description"`
}

func main() {

	// Connect to the MySQL database
	db, err := sql.Open("mysql", "root:connect@123@tcp(127.0.0.1:3306)/OES")
	if err != nil {
		log.Fatal("Failed to connect to the database:", err)
	}
	defer db.Close()

	// Ping the database to ensure the connection is established
	err = db.Ping()
	if err != nil {
		log.Fatal("Failed to ping the database:", err)
	}

	// Create a Gin router
	r := gin.Default()

	// Read operation - GET /menus
	r.GET("/", func(c *gin.Context) {
		rows, err := db.Query("SELECT id, name, url,description FROM Menu")
		if err != nil {
			log.Println("Failed to retrieve menus from the database:", err)
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to retrieve menus"})
			return
		}
		defer rows.Close()

		var menus []Menu
		for rows.Next() {
			var menu Menu
			err := rows.Scan(&menu.ID, &menu.Name, &menu.URL, &menu.Description)
			if err != nil {
				log.Println("Failed to scan menu row:", err)
				c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to retrieve menus"})
				return
			}
			menus = append(menus, menu)
		}

		c.JSON(http.StatusOK, menus)
	})

	// Run the server
	r.Run(":9000")
}

File: Comparison/js/CRUD/crud.js
const express = require('express');
const mysql = require('mysql2');

// Create a MySQL connection
const connection = mysql.createConnection({
    host: '127.0.0.1',
    user: 'root',
    password: 'connect@123',
    database: 'OES',
    insecureAuth: true
});

// Connect to the database
connection.connect(err => {
    if (err) {
        console.error('Error connecting to the database:', err);
        return;
    }
    console.log('Connected to the database');
});

// Create an Express app
const app = express();
app.use(express.json());

// Read operation - GET /menus
app.get('/', (req, res) => {
    connection.query('SELECT * FROM Menu', (err, results) => {
        if (err) {
            console.error('Error retrieving menus:', err);
            res.status(500).json({ error: 'Failed to retrieve menus' });
        } else {
            res.json(results);
        }
    });
});


// Start the server
const port = 3000;
app.listen(port, () => {
    console.log(`Server running on port ${port}`);
});
